yara-x-dump/src/serializer.rs

```diff
@@ -0,0 +1,573 @@
+pub(crate) trait Serializer {
+    /// Serialize a message
+    ///
+    /// # Arguments
+    ///
+    /// * `message`: The message to serialize
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Result<String, Error>` where the `String` is the serialized
+    /// message in specified format and the `Error` is any error that occurred
+    /// during serialization
+    ///
+    /// # Errors
+    ///
+    /// * `Error::ParsingJSONError`: If the message is not a valid JSON
+    /// * `Error::ParsingYAMLError`: If the message is not a valid YAML
+    /// * `Error::ParsingTOMLError`: If the message is not a valid TOML
+    /// * `Error::ParsingXMLError`: If the message is not a valid XML
+    fn serialize(&self, message: &str) -> Result<String, Error>;
+}
+
+/// Implement the trait for the JSON serializer
+impl Serializer for JsonSerializer {
+    fn serialize(&self, message: &str) -> Result<String, Error> {
+        let value = serde_json::from_str::<serde_json::Value>(message)?;
+        Ok(serde_json::to_string_pretty(&value)? + "\n")
+    }
+}
+
+/// Implement the trait for the YAML serializer
+impl Serializer for YamlSerializer {
+    fn serialize(&self, message: &str) -> Result<String, Error> {
+        let value = serde_json::from_str::<serde_yaml::Value>(message)?;
+        Ok(serde_yaml::to_string(&value)?)
+    }
+}
+
+/// Implement the trait for the TOML serializer
+impl Serializer for TomlSerializer {
+    fn serialize(&self, message: &str) -> Result<String, Error> {
+        let value = serde_json::from_str::<toml::Value>(message)?;
+        Ok(toml::to_string_pretty(&value)?)
+    }
+}
+
+/// Implement the trait for the XML serializer
+impl Serializer for XmlSerializer {
+    fn serialize(&self, message: &str) -> Result<String, Error> {
+        // Create a new XML builder and get the XML
+        let mut xml_builder = xml2json_rs::XmlConfig::new()
+            .rendering(xml2json_rs::Indentation::new(b' ', 2))
+            .decl(xml2json_rs::Declaration::new(
+                xml2json_rs::Version::XML10,
+                Some(xml2json_rs::Encoding::UTF8),
+                Some(true),
+            ))
+            .root_name("file")
+            .finalize();
+        let xml = xml_builder.build_from_json_string(message)? + "\n";
+        Ok(xml)
+    }
+}
+
+/// A function that returns a trait object based on the format
+///
+/// # Arguments
+///
+/// * `format`: The format to return the trait object for
+///
+/// # Returns
+///
+/// Returns a `Result<Box<dyn Serializer>, Error>` where the `Box<dyn
+/// Serializer>` is the trait object for the specified format and the `Error`
+/// is any error that occurred during the process
+///
+/// # Errors
+///
+/// * `Error::UnsupportedFormat`: If the format is unsupported
+pub(crate) fn get_serializer(
+    format: &str,
+) -> Result<Box<dyn Serializer>, Error> {
+    match format {
+        // Return a JSON serializer
+        "json" => Ok(Box::new(JsonSerializer)),
```

Comment:
For something that is inherently an enum, use an enum type. Avoid an APIs where the user can provide wrong input values. If you can catch an error at compile time is better that catching it at runtime. Instead of passing a `&str` with fixed values pass a type like:

```rust
enum Format {
    Json,
    Yaml,
    Toml,
}
```

But again, only JSON and YAML are really necessary.

Reply:
Thank your for this. I have used enums for output formats together with validating input using `ValueEnum` derive to handle parsing in a better way.